logins-sql/src/db.rs

```diff
@@ -0,0 +1,844 @@
+            // directly. See https://www.zetetic.net/sqlcipher/sqlcipher-api/#key
+            // "Raw Key Data" example. Note that this would be required to open
+            // existing iOS sqlcipher databases).
+            format!("PRAGMA key = '{}';", escape_string_for_pragma(key))
+        } else {
+            "".to_owned()
+        };
+
+        // `temp_store = 2` is required on Android to force the DB to keep temp
+        // files in memory, since on Android there's no tmp partition. See
+        // https://github.com/mozilla/mentat/issues/505. Ideally we'd only
+        // do this on Android, or allow caller to configure it.
+        let initial_pragmas = format!("
+            {}
+            PRAGMA temp_store = 2;
+        ", encryption_pragmas);
+
+        db.execute_batch(&initial_pragmas)?;
+
+        let mut res = Self { db };
+        schema::init(&mut res)?;
+        Ok(res)
+    }
+
+    pub fn open(path: impl AsRef<Path>, encryption_key: Option<&str>) -> Result<Self> {
+        Ok(Self::with_connection(Connection::open(path)?, encryption_key)?)
+    }
+
+    pub fn open_in_memory(encryption_key: Option<&str>) -> Result<Self> {
+        Ok(Self::with_connection(Connection::open_in_memory()?, encryption_key)?)
+    }
+
+    pub fn vacuum(&self) -> Result<()> {
+        self.execute("VACUUM")?;
+        Ok(())
+    }
+
+    pub fn execute_all(&self, stmts: &[&str]) -> Result<()> {
+        for sql in stmts {
+            self.execute(sql)?;
+        }
+        Ok(())
+    }
+
+    #[inline]
+    pub fn execute(&self, stmt: &str) -> Result<usize> {
+        Ok(self.do_exec(stmt, &[], false)?)
+    }
+
+    #[inline]
+    pub fn execute_cached(&self, stmt: &str) -> Result<usize> {
+        Ok(self.do_exec(stmt, &[], true)?)
+    }
+
+    #[inline]
+    pub fn execute_with_args(&self, stmt: &str, params: &[&ToSql]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, false)?)
+    }
+
+    #[inline]
+    pub fn execute_cached_with_args(&self, stmt: &str, params: &[&ToSql]) -> Result<usize> {
+        Ok(self.do_exec(stmt, params, true)?)
+    }
+
+    fn do_exec(&self, sql: &str, params: &[&ToSql], cache: bool) -> Result<usize> {
+        let res = if cache {
+            self.db.prepare_cached(sql)
+                   .and_then(|mut s| s.execute(params))
+        } else {
+            self.db.execute(sql, params)
+        };
+        if let Err(e) = &res {
+            warn!("Error running SQL {}. Statement: {:?}", e, sql);
+        }
+        Ok(res?)
+    }
+
+    #[inline]
+    pub fn execute_named(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec_named(stmt, params, false)?)
+    }
+
+    #[inline]
+    pub fn execute_named_cached(&self, stmt: &str, params: &[(&str, &ToSql)]) -> Result<usize> {
+        Ok(self.do_exec_named(stmt, params, true)?)
+    }
+
+    fn do_exec_named(&self, sql: &str, params: &[(&str, &ToSql)], cache: bool) -> Result<usize> {
```

Comment:
I see that you use the `_named` variants a lot (which makes the SQL easier to read than positional arguments!).  Can we just drop the un-`_named` variants?  Are those variants providing much value?

Reply:
IIRC the unnamed variants are still used for when we want to pass a variable number of e.g. guids in.